---
title: "pRDA candidates detection new variables"
author: "Thomas Francisco"
date: "2024-04-30"
output: html_document
---


```{r setup, include=FALSE}
rm(list = ls())
knitr::opts_chunk$set(
	message = FALSE,
	warning = FALSE,
	cache = FALSE
)
library(vegan)
library(dplyr)
library(robust) #for rdadapt
library(qvalue) #for rdadapt
library(ggplot2)
library(radiant.data) #for row_names_to_columns
library(textshape) #for colnames_to_row.names
```

**Correcting for population structure**  

pRDA is a version of the Redundancy analysis used to correct for some variables while performing an RDA. The steps are exactly the same as an pRDA but 1 step is performed before. Indeed, the goal of the pRDA is to perform an RDA  while correcting for some other variables that could have an impact on the responses or/and the explanatory variables.To account for that, pRDA first perform a linear regression between the response var and the explanatory var and with the conditioned variables (extra variables that could impact the response/ explanatory variables) as covariates. Then we performed the same step as an RDA but on the residual of the response variables and the explanatory variables because the residuals are the variance not explained by the conditioned variables -> this enable the model to remove the effects of the conditioned variables on the responses and explanatory variables.
Sources: Legendre and Legendre 2012, Capblancq 2021, workshop redundancy analysis of the Quebec centre for Biodiversity Science (https://r.qcbs.ca/workshop10/book-en/redundancy-analysis.html)


We will correct for population structure using only the 2 firsts PCa axis to avoid over correcting and to be consistent with other analysis (LFMM)


    1. DATA
    
Data is coming from the script variance_partitioning where the different variables of interest were gather in one data frame. 
The genomic matrix is the allelic frequencies at the population level (29 pop) on the imputed dataset (475 indiv, 8616 SNPs) with MAF correction
```{r data}

#genomic data
load("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/genetic_new/data_allelic_frequencies_29pop_adapcon_gentree_475_8616.Rdata")

genomic_matrix <- data_allelic_frequencies_29pop_adapcon_gentree_475_8616

#climatic/structure/IBD
load("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/GEA_new_var/variance_partitioning/Climatic_data_RDA_pRDA.Rdata") 
```

    2. Performing the RDA

We perform the pRDA accounting for population structure
```{r pRDA with population correction, message=FALSE, warning=FALSE}
#model
pRDA_env <- rda(formula = genomic_matrix ~  Annual_Tc+Diurnal_range_Tc+Tc_Seasonality+Tc_driest_quarter+Annual_P+P_Seasonality + Condition(PC1+PC2), data = Climatic_data_RDA_pRDA, scale=F)

RsquareAdj(pRDA_env)

#other analysis to test the model
#summary(pRDA_env)
#significativity of the model
#anova.cca(pRDA_env)
#test the significativity of the axis and the climatic variables using permutation
#anova.cca(pRDA_env, step = 1000, by = "axis")
#anova.cca(pRDA_env, step = 1000, by = "term")
```
The explained genomic variance by the climatic variables is less important than in the RDA because a part of the genomic variance is now attributed to the population structure and not the climatic variables. We still have a quite important variance explained suggesting that performing a candidate selection using GEA make sense. 

To perform the candidate selection on this models, we will follow procedure in Capblancq et al. 2021 and we will identified candidate based on their extremeness along a distribution of Mahalanobis distances estimated between each locus and the center of the RDA space using a certain number of axes (K). 

First, we need to select the number of ordination axis (RDA axis) that we will retains
```{r screeplot}
#screeplot
plot(pRDA_env$CCA$eig, option="screeplot")

#in hist
screeplot_pRDA<- screeplot(pRDA_env)

#explained variance along each RDA axis
pRDA_env$CCA$eig*100/sum(pRDA_env$CCA$eig) # variance of each axis*100 divided by the sum of all the explained variance for all axis
```

```{r save screeplot, include=FALSE}
 #save
png("C:/Users/tfrancisco/Documents/Thèse/Results/species/taxus/GEA_new_var/pRDA/screeplot_pRDA.png");screeplot_pRDA<- screeplot(pRDA_env);dev.off()
```

Interpretation: based on this, we will keep the 2 first RDA axis to perform the candidate detection

Then, we can plot the position of each populations in the RDA space. 
```{r population representation pRDA space}
#score along the 2 first RDA axis
score_climatic_var <- as.data.frame(scores(pRDA_env, choices=c(1:2), display="bp"))
Score_population <- data.frame(pRDA_env$CCA$u[,c(1,2)])

#meta_data
meta_data_pop <- read.csv("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/Populations/taxus_sample_29pop.csv",h=T,sep=";",dec=",")

#merge for country info
Score_population_bis <- rownames_to_column(Score_population,"Population")
score_with_country_info <- merge(Score_population_bis,meta_data_pop[,c(1,2)],"Population")
score_with_country_info$Country <- as.factor(score_with_country_info$Country)

#explained variance along each RDA axis
explained_variance <- data.frame(pRDA_env$CCA$eig)*100/sum(pRDA_env$CCA$eig)

explained_variance_round <- round(explained_variance$pRDA_env.CCA.eig,digits=1)

group_palette <- c("Bosnia"="orangered3", "France"="gold2","Germany"= "darkorchid3", "Greece"="navyblue", "Italy"="turquoise2", "Norway"="green3", "Slovakia"="blue", "Slovenia"="red", "Spain"="black", "Sweden"="gray", "Switzerland"="orange", "UK"="darkgreen")


##Biplot with populations and climatic variables along the 2 first RDA axis
biplot_populations <- ggplot() +
  geom_hline(yintercept = 0, linetype = "dashed", color = gray(0.80), size = 0.6) +
  geom_vline(xintercept = 0, linetype = "dashed", color = gray(0.80), size = 0.6) +
  geom_point(data = score_with_country_info, aes(x = RDA1 * 3, y = RDA2 * 3, colour = Country), size = 2, alpha = 0.8) +
  geom_segment(data = score_climatic_var, aes(xend = RDA1, yend = RDA2, x = 0, y = 0), colour = "black", size = 0.15, linetype = 1, arrow = arrow(length = unit(0.02, "npc"))) +
  geom_text(data = score_climatic_var, aes(x=1.1*RDA1, y=1.1*RDA2, label = row.names(score_climatic_var)), size = 2.5)+
  xlab(paste0("RDA 1 (",explained_variance_round[1],"%)")) + 
  ylab(paste0("RDA 2 (",explained_variance_round[2],"%)")) +
  ggtitle("Biplot pRDA Populations") +
  scale_color_manual(name = "Countries", values = group_palette, labels = levels(score_with_country_info$Country)) +
  theme_bw(base_size = 11) +
  theme(legend.position="right", panel.grid = element_blank(), strip.text = element_text(size=11),plot.title = element_text(hjust = 0.5,color = "Black",face="italic")) +
  labs(color = "Country")

 print(biplot_populations)
```

```{r save biplot indiv, include=FALSE}
#save
pdf("C:/Users/tfrancisco/Documents/Thèse/Results/species/taxus/GEA_new_var/pRDA/biplot_pRDA_populations.pdf");print(biplot_populations);dev.off()
```

Interpretation: We can see that these axis are more complicated to interpret than the ones from RDA. We can not see a clear geographic or altitudinal pattern. The first axis seems to separate populations based on their diurnal temperature variation, seasonality of the precipitation and temperature of the driest quarter. The second axis seems to discriminate the populations based on their temperature seasonality, diurnal range temperature and annual precipitation and temperature.


    3. Identification of outliers

The next step is to perform the genome scan procedure to calculate the Mahalanobis distance and then qvalues and pvalues in the same way as in the RDA script. 

```{r genome scan function}
rdadapt<-function(rda,K)
{
  zscores<-rda$CCA$v[,1:as.numeric(K)]
  resscale <- apply(zscores, 2, scale)
  resmaha <- covRob(resscale, distance = TRUE, na.action= na.omit, estim="pairwiseGK")$dist
  lambda <- median(resmaha)/qchisq(0.5,df=K)
  reschi2test <- pchisq(resmaha/lambda,K,lower.tail=FALSE)
  qval <- qvalue(reschi2test)
  q.values_rdadapt<-qval$qvalues
  return(data.frame(p.values=reschi2test, q.values=q.values_rdadapt))
}

#Perform the fucntion to calculate the mahalanobis distance and then pvalues/ qvalues

genome_scan_pRDA <- rdadapt(pRDA_env,2)

#plot the distribution of pvalues 
Histogram_of_Pvalues_pRDA<- hist(genome_scan_pRDA$p.values,
     main= "Histogram of pRDA P-values",
     xlab= "P-values")
```
```{r pval distrib, include=FALSE}
#save the histogram
 png(filename="C:/Users/tfrancisco/Documents/Thèse/Results/species/taxus/GEA_new_var/pRDA/Histogram_of_Pvalues_pRDA.png");Histogram_of_Pvalues_pRDA<- hist(genome_scan_pRDA$p.values,
     main= "Histogram of pRDA P-values",
     xlab= "P-values");dev.off()
```
Interpretation: Such as in the RDA, the distribution of the pvalues seems to follow what is expected. 

    4. Selection of a threshold
    
Candidate selection based on thresholds: 
```{r selection of outliers}
# qvalues < 0.05
thres_qvalues <- 0.05

outliers_qvalues_5perc_pRDA <- data.frame(Loci = colnames(genomic_matrix)[which(genome_scan_pRDA$q.values<thres_qvalues)], p.value = genome_scan_pRDA$q.values[which(genome_scan_pRDA$q.values<thres_qvalues)], contig = unlist(lapply(strsplit(colnames(genomic_matrix)[which(genome_scan_pRDA$q.values<thres_qvalues)], split = "_"), function(x) x[1])))

nrow(outliers_qvalues_5perc_pRDA)

# qvalues < 0.10
thres_qvalues <- 0.10

outliers_qvalues_10perc_pRDA <- data.frame(Loci = colnames(genomic_matrix)[which(genome_scan_pRDA$q.values<thres_qvalues)], p.value = genome_scan_pRDA$q.values[which(genome_scan_pRDA$q.values<thres_qvalues)], contig = unlist(lapply(strsplit(colnames(genomic_matrix)[which(genome_scan_pRDA$q.values<thres_qvalues)], split = "_"), function(x) x[1])))

nrow(outliers_qvalues_10perc_pRDA)

#threshold pvalues < 0.05 with Bonferonni correction
thres_pvalues <- 0.05/length(genome_scan_pRDA$p.values)

outliers_pvalues <- data.frame(Loci = colnames(genomic_matrix)[which(genome_scan_pRDA$p.values<thres_pvalues)], p.value = genome_scan_pRDA$p.values[which(genome_scan_pRDA$p.values<thres_pvalues)], contig = unlist(lapply(strsplit(colnames(genomic_matrix)[which(genome_scan_pRDA$p.values<thres_pvalues)], split = "_"), function(x) x[1])))

nrow(outliers_pvalues)

#top 1%
##add colnames
#genome_scan_pRDA$snp_names <- colnames(genomic_matrix)

#outliers_top1perc <- genome_scan_pRDA %>% 
#  arrange(p.values) %>%
#slice(1:(0.01*nrow(.))) %>%  
#  as.data.frame()


```

We also perform the outlier detection based on the loadings of loci along the pRDA axis to search for extreme loadings on each retained axis following Forester et al. (2018). 

```{r extremness along RDA axis}
load.rda <- scores(pRDA_env, choices=c(1:3), display="species")

hist(load.rda[,1], main="Loadings on RDA1")
hist(load.rda[,2], main="Loadings on RDA2")
```

```{r perform the detection }
#function to do it
outliers_function <- function(x,z){
  lims <- mean(x) + c(-1, 1) * z * sd(x)     # find loadings +/-z sd from mean loading     
  x[x < lims[1] | x > lims[2]]               # locus names in these tails
}

cand1 <- data.frame(outliers_function(load.rda[,1],3)) # 3
colnames(cand1) <- c("pvalues"); cand1$snp_names <- row.names(cand1)
cand2 <- data.frame(outliers_function(load.rda[,2],3)) # 51
colnames(cand2) <- c("pvalues"); cand2$snp_names <- row.names(cand2)

#names of the candidate
outliers_extremness <- rbind (cand1,cand2)


```


    5. Graphic representations of the candidate: 
    
One things that can be done when using RDA or pRDA as candidate selection methods is that we can also plot the candidates and neutral candidate on the pRDA (or RDA) space to visualize the position of SNPs from the climatic variables. 
```{r mahalanobis distance FDR 0.05 threshold}

score_loci <- as.data.frame(scores(pRDA_env, choices=c(1:2), display="species", scaling="none"))
score_loci_outliers <- data.frame(names = row.names(score_loci), score_loci)
score_loci_outliers$FDR5 <- "Neutral"
score_loci_outliers$FDR5[score_loci_outliers$names%in%outliers_qvalues_5perc_pRDA$Loci] <- "candidates FDR 5%"
score_loci_outliers$FDR5 <- factor(score_loci_outliers$FDR5, levels = c("Neutral", "candidates FDR 5%"))
score_loci_outliers <- score_loci_outliers[order(score_loci_outliers$FDR5),]

#Biplot with SNPs and climatic variables along the two first RDA axis. 
 biplot_outliers_pRDA<- ggplot() +
  geom_hline(yintercept=0, linetype="dashed", color = gray(.80), size=0.6) +
  geom_vline(xintercept=0, linetype="dashed", color = gray(.80), size=0.6) +
  geom_point(data = score_loci_outliers, aes(x=RDA1*15, y=RDA2*15,colour=FDR5), size = 1.4) +
  geom_segment(data = score_climatic_var, aes(xend=RDA1, yend=RDA2, x=0, y=0), colour="black", size=0.15, linetype=1, arrow=arrow(length = unit(0.02, "npc"))) +
  geom_text(data = score_climatic_var, aes(x=1.1*RDA1, y=1.1*RDA2, label = row.names(score_climatic_var)), size = 2.5, family = "Times") +
  xlab(paste0("RDA 1 (",explained_variance_round[1],"%)")) + 
  ylab(paste0("RDA 1 (",explained_variance_round[2],"%)")) +
  ggtitle("pRDA space: candidates FDR 5%") +
  guides(color=guide_legend(title="Locus type")) +
  scale_color_manual(values = c("lightblue", "#F9A242FF")) +
  theme_bw(base_size = 11) +
  theme(legend.position="right", panel.grid = element_blank(), strip.text = element_text(size=11),plot.title = element_text(hjust = 0.5,color = "Black",face="italic"))
 
print(biplot_outliers_pRDA)
```

```{r save biplot, include=FALSE}
#save
pdf("C:/Users/tfrancisco/Documents/Thèse/Results/species/taxus/GEA_new_var/pRDA/biplot_outliers_pRDA_FDR_5perc.pdf");print(biplot_outliers_pRDA);dev.off()
```


We can also represent the top candidates with FDR 5%, the 10% candidates and the presumably neutrals into the pRDA space
```{r mahalanobis distance FDR 0.10 threshold}

score_loci <- as.data.frame(scores(pRDA_env, choices=c(1:2), display="species", scaling="none"))
score_loci_outliers <- data.frame(names = row.names(score_loci), score_loci)
score_loci_outliers$FDR5_10 <- "Neutral"
score_loci_outliers$FDR5_10[score_loci_outliers$names%in%outliers_qvalues_10perc_pRDA$Loci] <- "candidates FDR 10%"
score_loci_outliers$FDR5_10[score_loci_outliers$names%in%outliers_qvalues_5perc_pRDA$Loci] <- "candidates FDR 5%"
score_loci_outliers$FDR5_10 <- factor(score_loci_outliers$FDR5_10, levels = c("Neutral","candidates FDR 10%","candidates FDR 5%"))
score_loci_outliers <- score_loci_outliers[order(score_loci_outliers$FDR5_10),]

#Biplot with SNPs and climatic variables along the two first RDA axis including the less conservative threshold
 biplot_outliers_pRDA_LC<- ggplot() +
  geom_hline(yintercept=0, linetype="dashed", color = gray(.80), size=0.6) +
  geom_vline(xintercept=0, linetype="dashed", color = gray(.80), size=0.6) +
  geom_point(data = score_loci_outliers, aes(x=RDA1*15, y=RDA2*15,colour=FDR5_10), size = 1.4) +
  geom_segment(data = score_climatic_var, aes(xend=RDA1, yend=RDA2, x=0, y=0), colour="black", size=0.15, linetype=1, arrow=arrow(length = unit(0.02, "npc"))) +
  geom_text(data = score_climatic_var, aes(x=1.1*RDA1, y=1.1*RDA2, label = row.names(score_climatic_var)), size = 2.5) +
  xlab(paste0("RDA 1 (",explained_variance_round[1],"%)")) + 
  ylab(paste0("RDA 2 (",explained_variance_round[2],"%)")) +
   ggtitle("pRDA space: candidate FDR 10%") +
  guides(color=guide_legend(title="Locus type")) +
  scale_color_manual(values = c("lightblue","darkgreen","orange")) +
  theme_bw(base_size = 11) +
  theme(legend.position="right", panel.grid = element_blank(), strip.text = element_text(size=11),plot.title = element_text(hjust = 0.5,color = "Black",face="italic"))
 
 print(biplot_outliers_pRDA_LC)
```

```{r save biplot all set SNPs, include=FALSE}
 #save
pdf("C:/Users/tfrancisco/Documents/Thèse/Results/species/taxus/GEA_new_var/pRDA/biplot_outliers_pRDA_FDR_5_10perc.pdf");print(biplot_outliers_pRDA_LC);dev.off()
```


    6. Overlapping between methods

We want to see if the candidate from extremness and the FDR 5% are similar: 
```{r overlapping of the methods}

colnames(outliers_qvalues_5perc_pRDA)= c( "snp_names","p.value","contig")
#overlapping extremeness 2.5 and top 1%
overlapping <- merge(outliers_extremness,outliers_qvalues_5perc_pRDA,"snp_names")
```
Interpretation: 152 out of the 161 candidate identified by extremeness are overlapping with the 225 of qvalues. 
 

We will use FDR as a threshold to identified snp.  
We save the candidate identified by FDR thresholds: 

      - FDR 5% candidate: 

```{r save FDR 5% candidate}
outliers_pRDA_FDR5perc_T_Adapcon_gentree <- outliers_qvalues_5perc_pRDA[,-3]

#write_xlsx(outliers_pRDA_FDR5perc_T_Adapcon_gentree,"C:/Users/tfrancisco/Documents/Thèse/Results/species/taxus/GEA_new_var/outliers/outliers_pRDA_FDR5perc_T_Adapcon_gentree.xlsx")
save(outliers_pRDA_FDR5perc_T_Adapcon_gentree, file="C:/Users/tfrancisco/Documents/Thèse/Results/species/taxus/GEA_new_var/outliers/outliers_pRDA_FDR5perc_T_Adapcon_gentree.Rdata")
```

      - FDR 10% candidate: 

```{r save FDR 10% candidate}
outliers_pRDA_FDR10perc_T_Adapcon_gentree <- outliers_qvalues_10perc_pRDA[,-3]

#write_xlsx(outliers_RDA_FDR10perc_T_Adapcon_gentree,"C:/Users/tfrancisco/Documents/Thèse/Results/species/taxus/GEA/outliers/outliers_RDA_FDR5perc_T_Adapcon_gentree.xlsx")
save(outliers_pRDA_FDR10perc_T_Adapcon_gentree, file="C:/Users/tfrancisco/Documents/Thèse/Results/species/taxus/GEA_new_var/outliers/outliers_pRDA_FDR10perc_T_Adapcon_gentree.Rdata")
```
